src/software/ai/navigator/path_planner/path_planner_test.cpp

```diff
@@ -0,0 +1,255 @@
+                          {std::cos(5 * M_PI / 3) * 0.2, std::sin(5 * M_PI / 3) * 0.2}),
+                      false),
+              },
+          .should_return_path = false}};
+
+template <typename PlannerT>
+std::pair<std::string, PathPlannerConstructor> name_and_constructor()
+{
+    return std::pair<std::string, PathPlannerConstructor>(
+        typeid(PlannerT).name(), []() { return std::make_unique<PlannerT>(); });
+}
+
+std::vector<std::pair<std::string, PathPlannerConstructor>>
+    path_planner_names_and_constructors = {
+        // add path planner constructors here
+        name_and_constructor<ThetaStarPathPlanner>(),
+        // uncomment this if you want some tests to fail
+        //        name_and_constructor<StraightLinePathPlanner>()
+};
+
+
+void validatePath(const Path &path, const Point &start, const Point &dest,
+                  const Rectangle &navigable_area, const std::vector<Obstacle> &obstacles)
+{
+    // check for zero length path
+    if (path.valueAt(0.f) == path.valueAt(1.f))
+    {
+        throw std::string("zero length path!");
+    }
+
+    // compute an s value interval that roughly corresponds to length
+    // PATH_CHECK_INTERVAL_M assuming that the path is locally linear
+    double s_per_meter_ratio   = 0.05 / (path.valueAt(0.05) - path.valueAt(0.0)).length();
+    double path_check_interval = s_per_meter_ratio * PATH_CHECK_INTERVAL_M;
+
+    std::cout << "Evaluating path at intervals of s=" << path_check_interval << std::endl;
+
+    for (double s = 0.0; s <= 1.0; s += path_check_interval)
+    {
+        Point pt = path.valueAt(s);
+        for (const Obstacle &obs : obstacles)
+        {
+            if (obs.containsPoint(pt))
+            {
+                // fail because path intersects obstacle
+                std::stringstream fail_ss;
+                fail_ss << "Point " << pt << "at s=" << s << " intersects obstacle {"
+                        << obs << "}";
+                throw fail_ss.str();
+            }
+        }
+        if (!navigable_area.contains(pt))
+        {
+            // fail because path exited navigable area
+            std::stringstream fail_ss;
+            fail_ss << "Point " << pt << "at s=" << s << " is not in navigable area "
+                    << navigable_area;
+            throw fail_ss.str();
+        }
+    }
+
+    // check that the path reaches the destination if the destination is not in an
+    // obstacle otherwise check that the path makes progress toward the destination i.e.
+    // the distance to dest is less at the end of the path than at the start of the path
+    bool dest_in_obstacle = std::any_of(
+        obstacles.begin(), obstacles.end(),
+        [&path](const auto &obs) { return obs.containsPoint(path.valueAt(1.0)); });
+
+    if (dest_in_obstacle &&
+        (path.valueAt(1.0) - dest).length() >= (path.valueAt(0.0) - dest).length())
+    {
+        // fail because no progress to destination
```

Comment:
Bit of a nitpick, but what if we had a testcase where a path started inside an obstacle, and the path also ended inside the same obstacle? The end point of the path would end up outside the obstacle, which would be correct, but the test would fail. I know this is somewhat of an edge case but I can see us writing a test like this.

Reply:
we definitely will need a test case where we start inside an obstacle and exit it, which is correct behavior. perhaps we could ignore obstacles that we start in? will explore this more tomorrow

Reply:
done